src/systems/physics-system.js

```diff
@@ -1,103 +1,311 @@
-        if (path.endsWith(".wasm")) {
-          return AmmoWasm;
-        }
-        return path;
-      }
-    });
+    this.ammoWorker = new AmmoWorker();
+    this.workerHelpers = new WorkerHelpers(this.ammoWorker);
 
-    this.bodies = [];
     this.bodyHelpers = [];
     this.shapeHelpers = [];
 
+    this.bodyUuids = [];
+    this.uuidToIndex = {};
+    this.indexToUuid = {};
+    this.object3Ds = {};
+    window.object3Ds = this.object3Ds;
+    this.bodyOptions = {};
+    this.bodyLinearVelocities = {};
+    this.bodyAngularVelocities = {};
+
+    this.shapeUuids = [];
+    this.shapes = {};
+
+    this.collisions = {};
+
+    this.constraints = {};
+
     this.debugRequested = false;
     this.debugEnabled = false;
     this.scene = scene;
     this.stepDuration = 0;
 
-    AmmoModule().then(() => {
-      this.world = new World(WORLD_CONFIG);
+    this.ready = false;
+    this.nextBodyUuid = 0;
+    this.nextShapeUuid = 0;
 
-      for (const bodyHelper of this.bodyHelpers) {
-        if (bodyHelper.alive) bodyHelper.init2();
-      }
-      for (const shapeHelper of this.shapeHelpers) {
-        if (shapeHelper.alive) shapeHelper.init2();
+    const arrayBuffer = new ArrayBuffer(4 * BUFFER_CONFIG.BODY_DATA_SIZE * BUFFER_CONFIG.MAX_BODIES);
+    this.objectMatricesFloatArray = new Float32Array(arrayBuffer);
+    this.objectMatricesIntArray = new Int32Array(arrayBuffer);
+
+    this.ammoWorker.postMessage(
+      {
+        type: MESSAGE_TYPES.INIT,
+        worldConfig: WORLD_CONFIG,
+        arrayBuffer,
+        wasmUrl: new URL(ammoWasmUrl, configs.BASE_ASSETS_PATH || window.location).href
+      },
+      [arrayBuffer]
+    );
+
+    this.ammoWorker.onmessage = async event => {
+      if (event.data.type === MESSAGE_TYPES.READY) {
+        this.ready = true;
+        for (const bodyHelper of this.bodyHelpers) {
+          if (bodyHelper.alive) bodyHelper.init2();
+        }
+        for (const shapeHelper of this.shapeHelpers) {
+          if (shapeHelper.alive) shapeHelper.init2();
+        }
+        this.shapeHelpers.length = 0;
+        this.bodyHelpers.length = 0;
+      } else if (event.data.type === MESSAGE_TYPES.BODY_READY) {
+        const uuid = event.data.uuid;
+        const index = event.data.index;
+        this.bodyUuids.push(uuid);
+        this.uuidToIndex[uuid] = index;
+        this.indexToUuid[index] = uuid;
+      } else if (event.data.type === MESSAGE_TYPES.SHAPES_READY) {
+        const bodyUuid = event.data.bodyUuid;
+        const shapesUuid = event.data.shapesUuid;
+        this.shapes[bodyUuid].push(shapesUuid);
+      } else if (event.data.type === MESSAGE_TYPES.TRANSFER_DATA) {
+        this.objectMatricesFloatArray = event.data.objectMatricesFloatArray;
+        this.objectMatricesIntArray = new Int32Array(this.objectMatricesFloatArray.buffer);
+        this.stepDuration = event.data.stepDuration;
       }
-      this.shapeHelpers.length = 0;
-      this.bodyHelpers.length = 0;
-    });
+    };
```

Comment:
Regarding the `onmessage` handler on the other side -- Do we expect messages to be coming in out-of-order, and if not should we be logging this as a bug? Surprised to see a strategy where we stick a message back into the queue if there wasn't a body it's associated with.
https://github.com/InfiniteLee/three-ammo/pull/2/files#diff-3db89a7579b555b025615250a93bb695R326